org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLTextDocumentService.java

```diff
@@ -395,39 +377,30 @@ private void triggerValidationFor(Collection<TextDocument> documents) {
 		if (!documents.isEmpty()) {
 			xmlLanguageServer.schedule(() -> {
 				documents.forEach(document -> {
-					String uri = document.getUri();
-					int version = document.getVersion();
-					doTriggerValidation(uri, version, monitor);
+					try {
+						CancelChecker cancelChecker = new TextDocumentVersionChecker(document, document.getVersion());
+						validate(document, cancelChecker);
+					} catch (CancellationException e) {
+						// Ignore the error and continue to validate other documents
+					}
 				});
 			}, 500, TimeUnit.MILLISECONDS);
 		}
 	}
 
-	private void triggerValidation(String uri, int version) {
-		if (future != null && !future.isCancelled()) {
-			future.cancel(true);
-		}
-		if (monitor != null) {
-			monitor.setCanceled(true);
-		}
-		monitor = new BasicCancelChecker();
-		triggerValidation(uri, version, monitor);
-	}
-
-	private void triggerValidation(String uri, int version, CancelChecker monitor) {
-		future = xmlLanguageServer.schedule(() -> {
-			doTriggerValidation(uri, version, monitor);
+	private void triggerValidationFor(TextDocument document) {
+		CancelChecker cancelChecker = new TextDocumentVersionChecker(document, document.getVersion());
+		xmlLanguageServer.schedule(() -> {
+			validate(document, cancelChecker);
 		}, 500, TimeUnit.MILLISECONDS);
 	}
 
-	private void doTriggerValidation(String uri, int version, CancelChecker monitor) {
-		TextDocument currDocument = getDocument(uri);
-		if (currDocument != null && currDocument.getVersion() == version) {
-			DOMDocument xmlDocument = getXMLDocument(currDocument);
-			getXMLLanguageService().publishDiagnostics(xmlDocument,
-					params -> xmlLanguageServer.getLanguageClient().publishDiagnostics(params),
-					(u, v) -> triggerValidation(u, v), monitor, sharedSettings.validationSettings);
-		}
+	private void validate(TextDocument document, CancelChecker cancelChecker) throws CancellationException {
+		cancelChecker.checkCanceled();
+		DOMDocument xmlDocument = getXMLDocument(document);
+		getXMLLanguageService().publishDiagnostics(xmlDocument,
+				params -> xmlLanguageServer.getLanguageClient().publishDiagnostics(params),
+				(doc) -> triggerValidationFor(doc), sharedSettings.validationSettings, cancelChecker);

```

Comment:
here, the client cancel checker is not used by the TextDocumentVersionChecker in triggerValidationFor, it should be wrapped in theory, but since validation is not something controlled by the client today, it's probably fine for now. Still, we should fix it in the next iteration